smoderp2d/core/surface.py

```diff
@@ -393,56 +339,47 @@ def rill_runoff(dt, effect_vrst, ratio, h_rill, rillWidth, v_rill_rest,
 
     q_rill = v_rill * h * b
 
-    vol_rill = q_rill * dt
-
-    courant = (v_rill * dt) / effect_vrst
-
-    # celerita
-    # courant = (1 + s*b/(3*(b+2*h))) * q_rill/(b*h)
-
-    v_rill_rest = ma.where(
-        courant <= courantMax,
-        ma.where(vol_rill > vol_to_rill, 0, vol_to_rill - vol_rill),
-        v_rill_rest
-    )
-    vol_runoff_rill = ma.where(
-        courant <= courantMax,
-        ma.where(vol_rill > vol_to_rill, vol_to_rill, vol_rill),
-        vol_runoff_rill
-    )
-
-    return (
-        v_rill, v_rill_rest, vol_runoff_rill, ratio, courant,
-        vol_to_rill, b
-    )
+    vol_runoff_rill = ma.filled(q_rill,0)
+        
+    return vol_runoff_rill
 
 
-def surface_retention(bil, sur):
-    """TODO.
+def surface_retention_impl(h_sur, reten_old):
+    reten = reten_old
+    # print(reten.max(), reten.min())
+    
+    h_ret = ma.where(reten<0, 
+                     ma.where(h_sur+reten > 0, reten, 
+                     -h_sur),
+                     0
+                     )
+    
+    return h_ret
 
-    :param bil: TODO
-    :param sur: TODO
-    """
+def surface_retention_update(h_sur, sur):
     reten = sur.sur_ret
-    pre_reten = reten
-    bil_new = ma.where(
-        reten < 0,
-        ma.where(bil + reten > 0, bil + reten, 0),
-        bil
-    )
     reten_new = ma.where(
         reten < 0,
-        ma.where(bil + reten > 0, 0, bil + reten),
+        ma.where(h_sur + reten > 0, 0, h_sur + reten),
         reten
     )
 
     sur.sur_ret = reten_new
-    sur.cur_sur_ret = reten_new - pre_reten
-
-    return bil_new
+    sur.cur_sur_ret = reten_new - reten
+    
+def inflows_comp(tot_flow, list_fd):
+    inflow = ma.array(ma.zeros((GridGlobals.r, GridGlobals.c)),mask=GridGlobals.masks)
+    r = GridGlobals.r
+    c = GridGlobals.c
+   
+    inflow[1:r,0:c-1] += list_fd[1:r,0:c-1,0]*tot_flow[0:r-1,1:c] #NE
+    inflow[1:r,0:c] += list_fd[1:r,0:c,1]*tot_flow[0:r-1,0:c] #N
+    inflow[1:r,1:c] += list_fd[1:r,1:c,2]*tot_flow[0:r-1,0:c-1] #NW
+    inflow[0:r,1:c] += list_fd[0:r,1:c,3]*tot_flow[0:r,0:c-1] #W
+    inflow[0:r-1,1:c] += list_fd[0:r-1,1:c,4]*tot_flow[1:r,0:c-1] #SW
+    inflow[0:r-1,0:c] += list_fd[0:r-1,0:c,5]*tot_flow[1:r,0:c] #S
+    inflow[0:r-1,0:c-1] += list_fd[0:r-1,0:c-1,6]*tot_flow[1:r,1:c] #SE
+    inflow[0:r,0:c-1] += list_fd[0:r,0:c-1,7]*tot_flow[0:r,1:c] #E
```

Comment:
Does this take into account the fact that some of these values might be masked? I have the feeling that these in-place operations modify even masked items (but I might be wrong, too...).

Reply:
I am not sure. Probably yes. We agreed that this will be solved in the future.